Enable expanding/collapsing error list items if with long message

yongyan-gh opened the pull request:
# Description

Enable error list row expando by setting its columns `StandardTableKeyNames.Text` and `StandardTableKeyNames.FullText` with short and full message texts separately
- When the original message contains line-breaker or end-of-sentence punctation chars in the middle.
- Extract first sentence as short message.
  - if first sentence does not end with a punctation, append a dot (.) at the end.
  - if first sentence length is greater than specific max length (165 in this case), it will be truncated to 165 and append a horizontal ellipsis char (…)
- The original message as full message.
- Support hyperlink in error list item and expanded panel.

Add test cases in Apex UI tests: https://github.com/microsoft/sarif-internal/issues/22

shaopeng-gh commented on src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:
```diff
@@ -94,8 +94,8 @@ public SarifErrorListItem(Run run, int runIndex, Result result, string logFilePa
 
-            this.RawMessage = result.GetMessageText(rule);
-            this.ShortMessage = this.RawMessage;
+            this.RawMessage = result.GetMessageText(rule, concise: false).Trim();
```
>.Trim() [](http://example.com/codeflow?start=73&length=7)

to whom has concern no null check before Trim, I have checked the GetMessageText() will never return null and return string.Empty instead

shaopeng-gh commented on src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:
```diff
@@ -94,8 +94,8 @@ public SarifErrorListItem(Run run, int runIndex, Result result, string logFilePa
 
-            this.RawMessage = result.GetMessageText(rule);
-            this.ShortMessage = this.RawMessage;
+            this.RawMessage = result.GetMessageText(rule, concise: false).Trim();
```
>Trim() [](http://example.com/codeflow?start=74&length=6)

open question, if the trim is needed

EasyRhinoMSFT replied:
This is for clean data, in the event that, for example, such characters cause some undesirable behavior in a future VS update

shaopeng-gh commented on src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifErrorListItemTests.cs:
```diff
@@ -388,7 +387,27 @@ public void SarifErrorListItem_ResultMessageFormat_NoTrailingPeriod()
             SarifErrorListItem item = MakeErrorListItem(result);
             item.Message.Should().Be(s1);
-            item.ShortMessage.Should().Be(s1);
+            item.ShortMessage.Should().Be($"{s1}.");
```
>item.ShortMessage.Should().Be($"{s1}."); [](http://example.com/codeflow?start=12&length=40)

is it by design we want to add a "."

shaopeng-gh replied:
I see we only adding "." to ShortMessage but not Message
Looks like not consistent.
ShortMessage should not be longer than Message

yongyan-gh replied:
You are right this is the behavior of existing Sarif-Sdk extension methods. We need our own methods to handle the differences.

yongyan-gh replied:
With a new implementation this is fixed

shaopeng-gh commented:
provide some description in the PR description

yongyan-gh commented on src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:
```diff
@@ -94,8 +94,8 @@ public SarifErrorListItem(Run run, int runIndex, Result result, string logFilePa
-            this.RawMessage = result.GetMessageText(rule);
-            this.ShortMessage = this.RawMessage;
+            this.RawMessage = result.GetMessageText(rule, concise: false).Trim();
+            this.ShortMessage = result.GetMessageText(rule, concise: true, maxLength: 165).Trim();
```
moved to a constant

yongyan-gh commented:
signed

---
In reply to: [1113092067](https://github.com/microsoft/sarif-visualstudio-extension/pull/481#issuecomment-1113092067) [](http://example.com/codeflow?ancestors=1113092067,1113092067)

yongyan-gh commented:
updated the description

---
In reply to: [1112541926](https://github.com/microsoft/sarif-visualstudio-extension/pull/481#issuecomment-1112541926) [](http://example.com/codeflow?ancestors=1112541926,1112541926)

EasyRhinoMSFT commented on src/Sarif.Viewer.VisualStudio.Core/SdkUiUtilities.cs:
```diff
@@ -701,6 +701,89 @@ internal static string GetPlainText(IEnumerable<XamlDoc.Inline> inlines)
+
+            const string ellipsis = "\u2026";
+            const string hyperlinkGroup = "link_text";
+            string pattern = string.Format(@"\[(?<{0}>[\w \.]+)\]\(([\w\.:\/ ]*)\)", hyperlinkGroup);
```
>{0} [](http://example.com/codeflow?start=50&length=3)

Can you use interpolation here? easier to read

EasyRhinoMSFT commented on src/Sarif.Viewer.VisualStudio.Core/SdkUiUtilities.cs:
```diff
@@ -701,6 +701,89 @@ internal static string GetPlainText(IEnumerable<XamlDoc.Inline> inlines)
+                {
+                    if (maxLength > 0 && match.Index + group.Length <= sb.Length)
+                    {
+                        string text = group.Value;
```
>string text = group.Value; [](http://example.com/codeflow?start=24&length=26)

Unused

EasyRhinoMSFT commented:
>     public class SarifErrorListItemTests

We should make sure there is coverage for messages that are close to the max length, and that have existing punctuation. Also make sure it handles maxLength strings that end with one or more spaces. You could write a single test method and then define a list of test cases (input strings with a corresponding expected output).

---
Refers to: src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifErrorListItemTests.cs:18 in 39828f5. [](commit_id = 39828f52759cf4369e2cf3e5b93d1ade5925d8bf, deletion_comment = False)

marmegh commented:
Don't forget the Release History. If the file doesn't already exist, please create one. This is definitely a bug fix.

Otherwise, I've tested locally with multiple link and location scenarios and everything looks great. Restored the expand/collapse feature without losing anything else in the process. I compared the same sarif examples between current and this version as part of validation. Once this PR is complete, I'll create a follow up PR to document testing.

yongyan-gh commented:
>     public class SarifErrorListItemTests

test cases added in SarifResultEntryTests

---
In reply to: [1113523845](https://github.com/microsoft/sarif-visualstudio-extension/pull/481#issuecomment-1113523845) [](http://example.com/codeflow?ancestors=1113523845,1113523845)

---
Refers to: src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifErrorListItemTests.cs:18 in 39828f5. [](commit_id = 39828f52759cf4369e2cf3e5b93d1ade5925d8bf, deletion_comment = False)

yongyan-gh commented:
Thanks, updated the release history, added missing entries I can find from all releases I can find (from **v3.0.16**)

---
In reply to: [1116329192](https://github.com/microsoft/sarif-visualstudio-extension/pull/481#issuecomment-1116329192) [](http://example.com/codeflow?ancestors=1116329192,1116329192)

EasyRhinoMSFT reviewed (approved):
:shipit:

marmegh commented on src/ReleaseHistory.md:
```diff
@@ -1,5 +1,33 @@
+
+## [SARIFER](https://marketplace.visualstudio.com/items?itemName=WDGIS.MicrosoftSarifer)
+* Publish official release to marketplace.
+* Enable Spam Analyzer Dynamic Mode.
```
SARIFER appears to have been released as v 3.0.73.27569. Could you please add the version number?

yongyan-gh replied:
Will update. I don't see this release at [https://github.com/microsoft/sarif-visualstudio-extension/releases](https://github.com/microsoft/sarif-visualstudio-extension/releases) where did you see it?

marmegh replied:
The VS marketplace.

shaopeng-gh reviewed (approved):
:shipit:
